src/Lykke.Job.CandleHistoryWriter.Repositories/HistoryMigration/HistoryProviders/TradesSQLHistory/TradesSqlHistoryRepository.cs

```diff
@@ -92,37 +95,27 @@ string searchToken
                     }
                 }
 
+                _sqlConnection.Close();
+
                 if (result.Count > 0)
                 {
                     // Now we need to remove the last several trades which have the same date and time (accuracy - to seconds).
                     // This will guarantee that we did not peek up some orders of the same trade on this iteration, and others
                     // on the next. On the next iteration we will read them again for the next batch. No temporary buffer, for
                     // it can't save any observable value of time. NOTE: if we have got less records than _sqlQueryBatchSize,
-                    // this means that we obtained the last )or the single) data pack, and there is no reason to delete "tail"
+                    // this means that we obtained the last (or the single) data pack, and there is no reason to delete "tail"
                     // trades.
 
                     if (result.Count == _sqlQueryBatchSize)
                     {
-                        // I do not want to use Linq.Remove(func) for it might case full sequence iterating which is slower than
-                        // the following messy code. Actually, we need to iterate only a bit of elements from the tail-side.
                         var lastDateTime = result.Last().DateTime.TruncateTo(CandleTimeInterval.Sec);
-                        var equalTimestampCandlesAtTailCount = 1;
-                        while (equalTimestampCandlesAtTailCount++ < result.Count) // Looking through the list starting from the last element. No additional work.
-                        {
-                            if (result[result.Count - equalTimestampCandlesAtTailCount].DateTime.TruncateTo(CandleTimeInterval.Sec) != lastDateTime)
-                            {
-                                equalTimestampCandlesAtTailCount--; // When we got the first inconsistent element, we need to step back.
-                                break;
-                            }
-                        }
+                        var resultWithoutTail = result.TakeWhile(t => t.DateTime < lastDateTime).ToList();
 
-                        // Now remove the candles with the same (to seconds) timestamp ONLY in case if we do have any other candles in the list.
-                        if (equalTimestampCandlesAtTailCount < result.Count)
-                            result.RemoveRange(result.Count - equalTimestampCandlesAtTailCount, equalTimestampCandlesAtTailCount);
+                        if (resultWithoutTail.Count != result.Count)
```

Comment:
This is connected with the previous comment: if we have a batch with equal dates (it will mean we've got the last trades in storage for our query), TakeWhile will return emptyness. But this is still a correct case 'cause the result itself contained something.
Additionally, if we've got an empty result of the query, it means "no data" which is a correct case too. Thus, an exception is not needed here.

Reply:
If we've got the last batch, then ```if (result.Count == _sqlQueryBatchSize)``` will avoid execution of the given code. And this became not a correct case. It means that entire batch contains trades with the same timestamp and all of them will be excluded and this will repeat on the next iteration and so on - program will stuck.

Reply:
> * if there are only trades with DateTime == lastDateTime, we return the result without changes;

Exactly this case is can lead to infine processing of the batches. In this case we should throw. That what i meant above. It seems that this case is impossible, but everything that could happen, will happen, and if it will happen, we should see it, instead of silently stuck.

And for the current code, you could remove this condition ```if (resultWithoutTail.Count != result.Count)``` (```result = resultWithoutTail;``` will be keeped) whitout any impact on code behaviour, because:

* if ```resultWithoutTail.Count != result.Count``` is true, then actually you return ```resultWithoutTail```
* if ```resultWithoutTail.Count != result.Count``` is false then ```result``` and ```resultWithoutTail``` are equals, and it's no matter what to return, so if you return ```resultWithoutTail``` in this case too, it will be ok.

So, you could do ```result = resultWithoutTail;``` unconditionally.